packages/core/src/rules/oas3/component-name-unique.ts

```diff
@@ -0,0 +1,157 @@
+import { Problem, UserContext } from '../../walk';
+import { Oas2Rule, Oas3Rule, Oas3Visitor } from '../../visitors';
+import {
+  Oas3Definition,
+  Oas3Parameter,
+  Oas3RequestBody,
+  Oas3Response,
+  Oas3Schema,
+  OasRef,
+} from '../../typings/openapi';
+
+const TYPE_NAME_SCHEMA = 'Schema';
+const TYPE_NAME_PARAMETER = 'Parameter';
+const TYPE_NAME_RESPONSE = 'Response';
+const TYPE_NAME_REQUEST_BODY = 'RequestBody';
+
+const TYPE_NAME_TO_OPTION_COMPONENT_NAME: { [key: string]: string } = {
+  [TYPE_NAME_SCHEMA]: 'schemas',
+  [TYPE_NAME_PARAMETER]: 'parameters',
+  [TYPE_NAME_RESPONSE]: 'responses',
+  [TYPE_NAME_REQUEST_BODY]: 'requestBodies',
+};
+
+export const ComponentNameUnique: Oas3Rule | Oas2Rule = (options) => {
+  const components = new Map<string, Set<string>>();
+
+  const typeNames: string[] = [];
+  if (options.schemas !== 'off') {
+    typeNames.push(TYPE_NAME_SCHEMA);
+  }
+  if (options.parameters !== 'off') {
+    typeNames.push(TYPE_NAME_PARAMETER);
+  }
+  if (options.responses !== 'off') {
+    typeNames.push(TYPE_NAME_RESPONSE);
+  }
+  if (options.requestBodies !== 'off') {
+    typeNames.push(TYPE_NAME_REQUEST_BODY);
+  }
+
+  function getOptionComponentNameForTypeName(typeName: string): string | null {
```

Comment:
Let's move function definitions after the return statement so they don't interrupt the code flow.

Reply:
Good idea. I moved them after the return statement. And those that don't depend on the `components` map I moved outside of the function to the end of the file.